Update BlockStore to index address hash -> transaction hashes

quantumagi opened the pull request:
Currently the Block Repository supports NoSQL indexes for "Block" and "Transaction".

- This pull request (#205) adds "Script" and "Output" indexes. 
- Index insertion has been changed to key order as that provides performance gains.

Indexing is dependent on the "-txindex" command line switch as with "Transaction" indexing. This can be changed to give us finer grained control.

Using the new "Script" index an address hash can be used to select the list of outputs (transaction hash + output number) of the transactions that the address appears in:

- Task<List<byte[]>> GetTrxOutByAddrAsync(BitcoinAddress addr)

Using the new "Outputs" index also allows us to determine which transactions these outputs have been spent to. Null entries will appear in the output list for corresponding unspent outputs:

- Task<List<uint256.>> GetTrxOutNextAsync(List<byte[]> trxOutList);

The functionality has not yet been exposed via RPC. Once this has been done it could assist with small-footprint mobile wallets, block explorers, etc.

Neurosploit commented on Stratis.Bitcoin/BlockStore/BlockRepository.cs:
```diff
@@ -193,15 +194,18 @@ public Task PutAsync(uint256 nextBlockHash, List<Block> blocks)
+                            foreach (var addrN in this.IndexedTransactionOutputs(transaction))
                             {
-                                HashSet<uint256> list = addrDict.TryGet(addr);
+                                var addr = (uint160)addrN[0];
```
Check for addrN length. Or throw exception if less than 2 long.

Neurosploit commented on Stratis.Bitcoin/BlockStore/BlockRepository.cs:
```diff
@@ -265,18 +269,19 @@ public Task PutAsync(uint256 nextBlockHash, List<Block> blocks)
-                    Array.Copy(script, 3, bytes, 0, 20);             
-                    yield return new uint160(bytes);
+                    Array.Copy(script, 3, bytes, 0, 20);
+                    yield return new object[2] { new uint160(bytes), N };
```
You could use a tuple here.

quantumagi replied:
Thank you. Fixed.

Neurosploit commented on Stratis.Bitcoin/BlockStore/BlockRepository.cs:
```diff
@@ -293,40 +453,203 @@ public Task DeleteAsync(uint256 newlockHash, List<uint256> hashes)
+                                    var value = trxId.ToBytes().Concat(((uint)addrN[1]).ToBytes());
+
+                                    HashSet<byte[]> list;
+                                    if (!removals.TryGetValue(addr, out list))
```
Did you know you can since C# 7 write this as if(!removals.TryGetValue(add, out HashSet<byte[]> list)) and skip the variable declaration on line 485?

quantumagi replied:
Thanks, I did not! I'm getting back into C# :D. I guess there is a slight trade-off between readability (being able to easily find the declaration) and being more concise.

Neurosploit commented on Stratis.Bitcoin/BlockStore/BlockRepository.cs:
```diff
@@ -293,40 +453,203 @@ public Task DeleteAsync(uint256 newlockHash, List<uint256> hashes)
+                cnt = BitConverter.ToUInt32(bytes, 0);
+            }
+
+            const int elementSize = 36; // Transaction Hash + output number = sizeof(trxid) + sizeof(uint)
```
If this is only a constant local to the method I would like to see it at the top of the method. If it's used elsewhere in the class put it above the constructor near the performancecounter.

quantumagi replied:
See below.

Neurosploit commented on Stratis.Bitcoin/BlockStore/BlockRepository.cs:
```diff
@@ -293,40 +453,203 @@ public Task DeleteAsync(uint256 newlockHash, List<uint256> hashes)
+
+        private int RemoveTransactionOutputsToAddress(uint160 addr, HashSet<byte[]> outputs)
+        {
+            const int elementSize = 36;
```
see previous comment. Merge into single const.

quantumagi replied:
I have added a small helper class that defines the constant TransactionOutputID.Length = 36.

Neurosploit commented on Stratis.Bitcoin/BlockStore/BlockRepository.cs:
```diff
@@ -293,40 +453,203 @@ public Task DeleteAsync(uint256 newlockHash, List<uint256> hashes)
+                        int j = 0;
+                        // trxid found?
+                        for (; j < elementSize; j++)
+                            if (listBytes[i + j] != output[j])
```
Can i+j go out of range of listbytes length? If so add a check.

quantumagi replied:
i <= listBytes.Length - TransactionOutputID.Length, and
j < TransactionOutputID.Length

As such i + j < listBytes.Length so it is guaranteed to remain in that range.

Also i >= 0 and j >= 0, so i + j >= 0.

dangershony commented:
Ah great I'll have a review tomorrow, I can already comment that I would expect this to be a standalone feature that depends on the BlockStore feature.
Or alternatively a replacement to the default BlockStore.

quantumagi commented:
I've updated the code the index scripts instead of addresses as it should be more useful in the long run. Currently there is still a filter in place to only index P2PKH scripts but this can be removed or made configurable.

dangershony commented on Stratis.Bitcoin/BlockStore/BlockRepository.cs:
```diff
@@ -139,58 +218,174 @@ public Task<uint256> GetTrxBlockIdAsync(uint256 trxid)
-				{
-					var blockId = block.GetHash();
+            // Use this comparer. We are assuming that DBreeze would use the same comparer for
+            // ordering rows in the file.
```
This is a very big assumption.

dangershony commented:
Ho @quantumagi so to start with I would say to build a separate Repository for index based on addresses.

Maybe call it AddressIndexRepository (ideally we can also pull out trx index but its build that way in core) 
Then you can create a new feature maybe call it IndexStoreFeature as we could potentially index many other things, it will have its own Signaled class for Blocks and its own RPC controllers.

Then when a node needs that capabilities they just need to add that feature.

```
  var node = new FullNodeBuilder()
   .UseNodeSettings(nodeSettings)
   .UseConsensus()
   .UseBlockStore()
   .UseMempool()
   .AddMining()
   .AddRPC()
   .UseIndexStore()
   .Build();

  node.Run();
```

Also to identify addresses use the template 
```
PayToPubkeyHashTemplate.Instance.ExtractScriptPubKeyParameters(scriptPubKey).ScriptPubKey.GetDestinationAddress(network)
```
And for pub keys
```
PayToPubkeyTemplate.Instance.ExtractScriptPubKeyParameters(scriptPubKey).Hash
```

And you could also add a flag to track P2SH

quantumagi commented:
@dangershony, thanks for pointing me to the templates.

A quick note that indexing script id's instead of addresses, as the code is currently doing, may be a good thing for future-proofing. Script id's can be created from addresses to key these look-ups, so we can even look up by address using the same index.

It may be useful to differentiate between "primary" indexes and other indexes. One way to look at it is that the primary indexes are unfiltered and nested in a sense:
- "Script" selects the outputs it appears in
- "Output" selects the transaction consuming it
- "Transaction" selects block
- "Block" selects block data

Going in the reverse direction should not require primary indexes as, block data selects block, block selects transactions, transaction selects outputs and output select script.

The above primary indexes could assist in creating derivative indexes (or filters), if required.

As for what to index, and where, it is apparent that you would also like to see it happen in a more consistent manner and not spread out over repositories and disparate commits. The blocks themselves are a bit of dilemma as they are a combination of block data and the block indexes and can't be cleanly separated.

This led to my initial approach of integration with the block repository's block and transaction indexing for single-source-of-truth. This seemed to offer a simple solution towards primary indexing consistency at database-commit-level.

However taking your suggestions on board another approach comes to mind:

- Have the indexer feature be responsible for activating additional indexing inside of the existing block store indexer. This will ensure that all primary indexes are updated together during the block commit. This can either be done by adding code to the existing block store indexer or by callbacks provided by the IndexStore feature.

dangershony commented:
Ah sorry forgot to answer.

I totally get the single-source-of-truth comment however that may not be necessary if the IndexStore is self contained (it can recover by itself and pull blocks from BlockStore if it needs).
Yeah indeed it be best to have a single commit to all pushed data, if thats the case perhaps extending the BlockStore is the better approach (closely to what you suggested)

The IndexStore will have a BlockStore inside plus other data, for that you could expose some of the BlockStore methods as virtual (or add an OnPersist virtual method)

Also remember we have the SingleThread requirement to think about.

quantumagi commented:
As per your comments above:
- "..extending the BlockStore is the better approach"
- "..expose some of the BlockStore methods as virtual"

Some of the change I made:
  - Added "protected virtual void OnInsertBlocks(List<Block> blocks)"
  - Added "protected virtual void OnInsertTransactions(List<(Transaction, Block)> transactions)"
  - Added "protected virtual void OnDeleteTransactions(List<Transaction> transactions)"
  - Added "protected virtual void OnDeleteBlocks(List<Block> blocks)"

dangershony commented on Stratis.Bitcoin/BlockStore/BlockRepository.cs:
```diff
@@ -139,58 +141,93 @@ public Task<uint256> GetTrxBlockIdAsync(uint256 trxid)
-                    {
-                        this.PerformanceCounter.AddRepositoryHitCount(1);
-                    }
+            var blockDict = new Dictionary<uint256, Block>();
```
Dictionary<byte[], Block>() will avoid the double ToByte() work (sort and serialize).

dangershony commented on Stratis.Bitcoin/BlockStore/BlockRepository.cs:
```diff
@@ -139,58 +141,93 @@ public Task<uint256> GetTrxBlockIdAsync(uint256 trxid)
+                OnInsertTransactions(transactions);
+        }
+
+        virtual protected void OnInsertTransactions(List<(Transaction, Block)> transactions)
```
Do we need an OnInsertTransactions one can always use the OnInsertBlock
But for readability perhaps this is better.

dangershony commented on Stratis.Bitcoin/BlockStore/BlockRepository.cs:
```diff
@@ -139,58 +141,93 @@ public Task<uint256> GetTrxBlockIdAsync(uint256 trxid)
+            var byteListComparer = new ByteListComparer();
 
-		private bool? LoadTxIndex()
+            transactions.Sort((pair1, pair2) => byteListComparer.Compare(pair1.Item1.GetHash().ToBytes(), pair2.Item1.GetHash().ToBytes()));
```
Something to bear in mind is the sorting of an external list (transactions).
It would be safer to do a ToList() and sort the new list or use the OrderBy() linq extension 
Not a big deal though.

dangershony commented:
Excellent job!
You want to fix my comments? I am fine pushing and you fix it in your next PR.

saleemrashid commented:
@quantumagi Will a method to get transactions for a script between two block heights be implemented? We need it for the TREZOR integration
